plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+  where
+    -- | Float bindings in the given `Term` inwards, and calculate the set of
+    -- variable `Unique`s in the result `Term`.
+    go ::
+        Term tyname name uni fun a ->
+        Term tyname name uni fun (a, Uniques)
+    go t = case t of
+        Apply a fun0 arg0 ->
+            let fun = go fun0
+                arg = go arg0
+                us = termUniqs fun `Set.union` termUniqs arg
+             in Apply (a, us) fun arg
+        LamAbs a n ty body0 ->
+            let body = go body0
+             in LamAbs (a, termUniqs body) n (noUniq ty) body
+        TyAbs a n k body0 ->
+            let body = go body0
+             in TyAbs (a, termUniqs body) n (noUniq k) body
+        TyInst a body0 ty ->
+            let body = go body0
+             in TyInst (a, termUniqs body) body (noUniq ty)
+        IWrap a ty1 ty2 body0 ->
+            let body = go body0
+             in IWrap (a, termUniqs body) (noUniq ty1) (noUniq ty2) body
+        Unwrap a body0 ->
+            let body = go body0
+             in Unwrap (a, termUniqs body) body
+        Let a NonRec bs0 body0 ->
+            let bs = go' <$> bs0
+                body = go body0
+             in -- The bindings in `bs` should be processed from right to left, since
+                -- a binding may depend on another binding to its left.
+                foldr (floatAlg ver a) body bs
+        Let a Rec bs0 body0 ->
+            -- Currently we don't move recursive bindings, so we simply descend into the body.
+            let bs = go' <$> bs0
+                body = go body0
+                us = Set.union (termUniqs body) (foldMap bindingUniqs bs)
+             in Let (a, us) Rec bs body
+        Var a n -> Var (a, Set.singleton (n ^. PLC.unique)) n
+        Constant{} -> (,mempty) <$> t
+        Builtin{} -> (,mempty) <$> t
+        Error{} -> (,mempty) <$> t
+
+    -- | Float bindings in the given `Binding` inwards, and calculate the set of
+    -- variable `Unique`s in the result `Binding`.
+    go' ::
+        Binding tyname name uni fun a ->
+        Binding tyname name uni fun (a, Uniques)
+    go' = \case
+        TermBind a s var rhs ->
+            let rhs' = go rhs
+             in TermBind (a, termUniqs rhs') s (noUniq var) rhs'
+        other -> noUniq other
+
+termUniqs :: Term tyname name uni fun (a, Uniques) -> Uniques
+termUniqs = snd . termAnn
+
+-- | The result includes both the `Unique` of the bound var and the `Unique`s in the RHS.
+bindingUniqs ::
+    PLC.HasUnique name PLC.TermUnique =>
+    Binding tyname name uni fun (a, Uniques) ->
+    Uniques
+bindingUniqs b = Set.union (snd (bindingAnn b)) $ case b of
+    TermBind _a _s var _rhs -> Set.singleton $ var ^. PLC.varDeclName . PLC.unique
+    _                       -> mempty
+
+noUniq :: Functor f => f a -> f (a, Uniques)
+noUniq = fmap (,mempty)
+
+{- | The algebra for floating a `Binding` inwards. Given a `Term` and a `Binding`, determine
```

Comment:
I've noticed you like to call things "algebras". Is this _actually_ an algebra, though? I'm not sure that it is. What about "floatInBinding" - that's what it does, no?

Reply:
It is an algebra. `foldr` is basically a catamorphism that turns an F-algebra into a fold.

To be precise, the `a -> b -> b` argument of `foldr` is only a part of the algebra. The full algebra is `(a -> b -> b, b)` (isomorphic to `ListF a b -> b`). But technically it is itself also an algebra.

Reply:
I guess I'm not sure I find that actually informative about what it does.